src/transformers/models/mot/modeling_mot.py

```diff
@@ -0,0 +1,1621 @@
+# coding=utf-8
+# Copyright 2024 The OpenAI Team Authors and HuggingFace Inc. team.
+# Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""PyTorch MixtureOfTokens model."""
+
+import math
+import warnings
+from typing import Optional, Tuple, Union
+
+import torch
+import torch.nn.functional as F
+import torch.utils.checkpoint
+from torch import nn
+from torch.cuda.amp import autocast
+from torch.nn import BCEWithLogitsLoss, CrossEntropyLoss, MSELoss
+from torch.nn.init import trunc_normal_
+
+from ...activations import ACT2FN
+from ...modeling_outputs import (
+    BaseModelOutputWithPastAndCrossAttentions,
+    CausalLMOutputWithCrossAttentions,
+    QuestionAnsweringModelOutput,
+    SequenceClassifierOutputWithPast,
+    TokenClassifierOutput,
+)
+from ...modeling_utils import PreTrainedModel
+from ...pytorch_utils import Conv1D, find_pruneable_heads_and_indices, prune_conv1d_layer
+from ...utils import (
+    add_code_sample_docstrings,
+    add_start_docstrings,
+    add_start_docstrings_to_model_forward,
+    logging,
+)
+from ...utils.model_parallel_utils import assert_device_map, get_device_map
+from .configuration_mot import MoTConfig
+
+
+logger = logging.get_logger(__name__)
+
+_CHECKPOINT_FOR_DOC = "jaszczur/mixture_of_tokens"
+_CONFIG_FOR_DOC = "MoTConfig"
+
+
+def with_batch_size_alignment(forward_fn):
+    def _forward(self, x):
+        """assumed ordering (batch, seq_len, dmodel)"""
+        size = x.size(self.sparsity_dim)
+        if size % self.group_size != 0:
+            if self.sparsity_dim == 1:
+                x = x.transpose(0, 1)
+
+            x = self.pad(x)
+
+            if self.sparsity_dim == 1:
+                x = forward_fn(self, x.transpose(0, 1))
+                return x[:, :size, :]
+            else:
+                x = forward_fn(self, x)
+                return x[:size, :, :]
+        else:
+            return forward_fn(self, x)
+
+    return _forward
+
+
+# Copied from transformers.models.gpt2.modeling_gpt2.GPT2Attention with GPT2->MoT
```

Comment:
Arf, GPT2 is a super old model, most probably not a good reference implementation. Lots of issues with the conv1d have been reported 😢